shared/lib_utility_rate_equations.cpp

```diff
@@ -215,37 +219,64 @@ bool rate_data::check_for_kwh_per_kw_rate(int units) {
 
 double rate_data::get_billing_demand(int month) {
     int m = 0;
-    double billing_demand = ec_bd_minimum;
-    int prev_yr_lookback = 11 - (ec_bd_lookback_months - month); // What month do we stop looking back in the prev yr?
+    double billing_demand = bd_minimum;
+    int prev_yr_lookback = 11 - (bd_lookback_months - month); // What month do we stop looking back in the prev yr?
 
     for (m = 11; m >= prev_yr_lookback && m >= 0; m--) {
-        double ratchet_percent = ec_bd_lookback_percents[m] * 0.01;
+        double ratchet_percent = bd_lookback_percents[m] * 0.01;
         double months_demand = prev_peak_demand[m] * ratchet_percent;
         if (months_demand > billing_demand) {
             billing_demand = months_demand;
         }
     }
 
     int start_month = 0;
-    if (month >= ec_bd_lookback_months) {
-        start_month = month - ec_bd_lookback_months;
+    if (month >= bd_lookback_months) {
+        start_month = month - bd_lookback_months;
     }
 
+    int idx = 0;
     for (m = start_month; m <= month; m++) {
-        double ratchet_percent = ec_bd_lookback_percents[m] * 0.01;
-        double months_demand = m_month[m].dc_flat_peak * ratchet_percent;
-        if (months_demand > billing_demand) {
-            billing_demand = months_demand;
+        idx = 0;
+        for (int p : m_month[m].dc_periods) {
+            if (bd_tou_periods.at(p)) {
+                double ratchet_percent = bd_lookback_percents[m] * 0.01;
+                double months_demand = m_month[m].dc_tou_peak[idx] * ratchet_percent;
+                if (months_demand > billing_demand) {
+                    billing_demand = months_demand;
+                }
+            }
+            idx++;
         }
     }
 
-    if (m_month[month].dc_flat_peak > billing_demand && m_month[month].use_current_month_ratchet) {
-        billing_demand = m_month[month].dc_flat_peak;
+    if (m_month[month].use_current_month_ratchet) {
+        idx = 0;
+        for (int p : m_month[month].dc_periods) {
+            if (bd_tou_periods.at(p)) {
+                double months_demand = m_month[month].dc_tou_peak[idx];
+                if (months_demand > billing_demand) {
+                    billing_demand = months_demand;
+                }
+            }
+            idx++;
+        }
     }
 
     return billing_demand;
 }
 
+void rate_data::set_billing_demands() {
+    for (int m = 0; m < (int) m_month.size(); m++) {
+        double flat_peak = m_month[m].dc_flat_peak;
+        if (en_billing_demand_lookback) {
+            // If ratchets are present the peak used here might be the actual peak, or something based on a previous month.
+            flat_peak = get_billing_demand(m);
```

Comment:
Is there a reference for this? i.e. are all flat demand charges to be based on the billing demand lookback? Seems like that overwrites the flat demand charge...

Reply:
Okay, you are not overwriting the ds_flat_peak - please disregard above comment.